flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ExecutionDecorator.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.streaming.runtime.tasks.mailbox.Mail;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Wraps execution of a {@link Runnable}, {@link ThrowingRunnable}, {@link Callable}, or {@link Mail}.
+ * Intended to customize execution in sub-types fo {@link org.apache.flink.streaming.runtime.tasks.StreamTask StreamTask},
+ * e.g. synchronization in {@link org.apache.flink.streaming.runtime.tasks.SourceStreamTask SourceStreamTask}.
+ */
+@Internal
+public interface ExecutionDecorator {
+	void run(Runnable runnable);
+
+	<E extends Throwable> void runThrowing(ThrowingRunnable<E> runnable) throws E;
+
+	<R> R call(Callable<R> callable) throws Exception;
+
+	void dispatch(Mail mail);
+
+	ExecutionDecorator NO_OP = new ExecutionDecorator() {
+		@Override
+		public void run(Runnable runnable) {
+			runnable.run();
+		}
+
+		@Override
+		public <E extends Throwable> void runThrowing(ThrowingRunnable<E> runnable) throws E {
+			runnable.run();
+		}
+
+		@Override
+		public <R> R call(Callable<R> callable) throws Exception {
+			return callable.call();
+		}
+
+		@Override
+		public void dispatch(Mail mail) {
+			mail.run();
+		}
+	};
+
+	/**
+	 * Returns an ExecutionDecorator that synchronizes each invocation.
+	 */
+	static SynchronizedExecutionDecorator syncExecutionDecorator() {
+		return syncExecutionDecorator(new Object());
+	}
+
+	/**
+	 * Returns an ExecutionDecorator that synchronizes each invocation on a given object.
+	 */
+	static SynchronizedExecutionDecorator syncExecutionDecorator(Object mutex) {
+		return new SynchronizedExecutionDecorator(mutex);
+	}
+
+	/**
+	 * A {@link ExecutionDecorator} that synchronizes every operation on the provided mutex.
+	 */
+	class SynchronizedExecutionDecorator implements ExecutionDecorator {
```

Comment:
On the other hand, `checkpointLock` is confusing for newcomers:) As it is used not only for checkpoints.
I think we can just add Javadoc.

Reply:
Yes, `checkpointLock` is quite confusing, but `mutex` is the most technically term out there and doesn't add any semantics. So I'm not happy with either solution. 
Since the whole class is about coordination of potentially parallel actions, I could live with `mutex`. But then we should definitively have `checkpointLock` in the first commit message line.

Reply:
We still have `getCheckpointLock()` method in `StreamTask` and `private Object checkpointLock` or  `private Object lock` in lots of places. 
The only change this PR introduces here is that `getCheckpointLock()` now delegates to the other class.

What do you mean by techincally term?
`mutex` field provides **mut**ual **ex**clusion for some sections of code. `lock` can be for example shared. A little bit more context.

Reply:
With added javadoc I'm fine with the current state.